lib/modules/manager/pub/artifacts.spec.ts

```diff
@@ -0,0 +1,230 @@
+import { GlobalConfig } from '../../../config/global';
+import type { RepoGlobalConfig } from '../../../config/types';
+import type { UpdateArtifactsConfig } from '../types';
+import * as pub from './artifacts';
+
+jest.mock('../../../util/exec/env');
+jest.mock('../../../util/fs');
+jest.mock('../../../util/git');
+jest.mock('../../../util/http');
+
+const adminConfig: RepoGlobalConfig = {
+  localDir: join('/tmp/github/some/repo'),
+  cacheDir: join('/tmp/cache'),
+  containerbaseDir: join('/tmp/cache/containerbase'),
+};
+
+const config: UpdateArtifactsConfig = {};
+
+describe('modules/manager/pub/artifacts', () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.resetModules();
+
+    env.getChildProcessEnv.mockReturnValue(envMock.basic);
+    GlobalConfig.set(adminConfig);
+  });
+
+  afterEach(() => {
+    GlobalConfig.set(adminConfig);
+  });
+
+  it('returns null if no pubspec.lock found', async () => {
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps,
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+  });
+
+  it('returns null if updatedDeps is empty', async () => {
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps: [],
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+  });
+
+  it('returns null if unchanged', async () => {
+    const execSnapshots = mockExecAll();
+    fs.readLocalFile.mockResolvedValueOnce('Current pubspec.lock');
+    fs.readLocalFile.mockResolvedValueOnce('Current pubspec.lock');
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps,
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+    expect(execSnapshots).toMatchObject([
+      {
+        cmd: 'dart pub upgrade',
+        options: {
+          cwd: '/tmp/github/some/repo',
+          encoding: 'utf-8',
+          env: {
+            HOME: '/home/user',
+            HTTPS_PROXY: 'https://example.com',
+            HTTP_PROXY: 'http://example.com',
+            LANG: 'en_US.UTF-8',
+            LC_ALL: 'en_US',
+            NO_PROXY: 'localhost',
+            PATH: '/tmp/path',
+          },
+          maxBuffer: 10485760,
+          timeout: 900000,
+        },
+      },
+    ]);
+  });
+
+  it('returns updated dart pubspec.lock', async () => {
+    const execSnapshots = mockExecAll();
+    fs.readLocalFile.mockResolvedValueOnce('Old pubspec.lock');
+    fs.readLocalFile.mockResolvedValueOnce('New pubspec.lock');
+    const updatedDeps = [{ depName: 'dep1' }];
+    expect(
+      await pub.updateArtifacts({
+        packageFileName: 'pubspec.yaml',
+        updatedDeps,
+        newPackageFileContent: '',
+        config,
+      })
+    ).not.toBeNull();
```

Comment:
should now here. use `toEqual` instead of `not.toBeNull` for all tests.

because it can resolve with artifacts error instead of success and vice versa.
so it can silently break with current solution.

Reply:
i still don't get it how i can use it without jest complaining. see my comment in above.

Reply:
```suggestion
    ).Equal([
      {
        file: {
          type: 'addition',
          path: 'pubspec.lock',
          contents: 'New pubspec.lock',
        },
      },
    ]);
```

Should be like that. do it for all `.not.toBeNull()` tests.